### PR TITLE
ARROW-9851: [C++] Disable AVX512 runtime paths with Valgrind

### PR DESCRIPTION
**Pull request by pitrou**
Introduce a CMake option to limit the maximum dynamic dispatch SIMD level, as Valgrind is still lacking support for AVX512 instructions (https://bugs.kde.org/show_bug.cgi?id=383010).

Also fix an uninitialized memory read due to GTest.

### CODE REVIEW

**Comment by pitrou** [comment]
@github-actions crossbow submit conda-cpp-valgrind

**Comment by pitrou** [comment]
@crossbow submit conda-cpp-valgrind

**Comment by pitrou** [comment]
@ursabot crossbow submit conda-cpp-valgrind

**Comment by pitrou** [comment]
@ursabot crossbow submit test-conda-cpp-valgrind

**Comment by pitrou** [comment]
@ursabot crossbow submit test-conda-cpp-valgrind

**Comment by pitrou** [comment]
@ursabot crossbow submit test-conda-cpp-valgrind

**Comment by pitrou** [comment]
@ursabot crossbow submit test-conda-cpp-valgrind

**Comment by pitrou** [comment]
@jianxind What is your opinion on the approach here?

**Comment by jianxind** [comment]
> @jianxind What is your opinion on the approach here?

It's fine to introduce ARROW_RUNTIME_SIMD_LEVEL flag which give more user options. Another potential approach is the ARROW_USER_SIMD_LEVEL flag during runtime, something like "ARROW_USER_SIMD_LEVEL=avx2 make unittest", I'm not familiar with CI system thus don't know if it's easy to hook ARROW_USER_SIMD_LEVEL flag.

+1

File: cpp/src/arrow/util/decimal_test.cc [hunk 1]
```diff
@@ -384,10 +389,10 @@ static const ToStringTestData kToStringTestData[] = {
     {-1234567890123456789LL, 25, "-1.234567890123456789E-7"},
 };
 
-class Decimal128ToStringTest : public ::testing::TestWithParam<ToStringTestData> {};
+class Decimal128ToStringTest : public ::testing::TestWithParam<ToStringTestParam> {};
 
 TEST_P(Decimal128ToStringTest, ToString) {
-  const ToStringTestData& data = GetParam();
+  const ToStringTestParam& data = GetParam();
```
Comment (kou) [inline cpp/src/arrow/util/decimal_test.cc:395]: How about changing local variable name to `param` from `data` too?

**Review by kou — APPROVED** [review]
+1

**Comment by kou** [comment]
@github-actions crossbow submit test-conda-cpp-valgrind

**Comment by kou** [comment]
> Another potential approach is the ARROW_USER_SIMD_LEVEL flag during runtime, something like "ARROW_USER_SIMD_LEVEL=avx2 make unittest", I'm not familiar with CI system thus don't know if it's easy to hook ARROW_USER_SIMD_LEVEL flag.

We can change environment variables only for test in CI.

But I don't know whether Valgrind can work with binary that has AVX512 instructions or not...

**Comment by jianxind** [comment]
> > Another potential approach is the ARROW_USER_SIMD_LEVEL flag during runtime, something like "ARROW_USER_SIMD_LEVEL=avx2 make unittest", I'm not familiar with CI system thus don't know if it's easy to hook ARROW_USER_SIMD_LEVEL flag.
> 
> We can change environment variables only for test in CI.
> 
> But I don't know whether Valgrind can work with binary that has AVX512 instructions or not...

Probably yes I think, as current approach only disable the AVX512 register path, the AVX512 file is still built into the lib.

The AVX512 file is defined under CXX_SUPPORTS_AVX512 flag
```
if(CXX_SUPPORTS_AVX512)
  list(APPEND ARROW_SRCS util/bpacking_avx512.cc)
  set_source_files_properties(util/bpacking_avx512.cc PROPERTIES SKIP_PRECOMPILE_HEADERS
                              ON)
  set_source_files_properties(util/bpacking_avx512.cc PROPERTIES COMPILE_FLAGS
                              ${ARROW_AVX512_FLAG})
endif()
```

**Comment by kou** [comment]
OK. Then I think that the environment variable approach is better because we already have `ARROW_USER_SIMD_LEVEL` support (I didn't know!). The following patch will work:

```diff
diff --git a/docker-compose.yml b/docker-compose.yml
index 86cffad81..55da4b60b 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -199,6 +199,7 @@ services:
       ARROW_JEMALLOC: "OFF"
       ARROW_S3: "OFF"
       ARROW_TEST_MEMCHECK: "ON"
+      ARROW_USER_SIMD_LEVEL: "AVX2" # AVX512 not supported by Valgrind (ARROW-9851)
       ARROW_USE_LD_GOLD: "ON"
       BUILD_WARNING_LEVEL: "PRODUCTION"
     volumes: *conda-volumes
```

**Comment by jianxind** [comment]
> OK. Then I think that the environment variable approach is better because we already have `ARROW_USER_SIMD_LEVEL` support (I didn't know!). The following patch will work:
> 
> ```diff
> diff --git a/docker-compose.yml b/docker-compose.yml
> index 86cffad81..55da4b60b 100644
> --- a/docker-compose.yml
> +++ b/docker-compose.yml
> @@ -199,6 +199,7 @@ services:
>        ARROW_JEMALLOC: "OFF"
>        ARROW_S3: "OFF"
>        ARROW_TEST_MEMCHECK: "ON"
> +      ARROW_USER_SIMD_LEVEL: "AVX2" # AVX512 not supported by Valgrind (ARROW-9851)
>        ARROW_USE_LD_GOLD: "ON"
>        BUILD_WARNING_LEVEL: "PRODUCTION"
>      volumes: *conda-volumes
> ```

https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/cpu_info.cc#L423

**Comment by kou** [comment]
> ARROW_USER_SIMD_LEVEL is a ENV variants, will this change pass AVX2 to ARROW_USER_SIMD_LEVEL env?

Yes. But the patch was wrong. We need to use lower case for `ARROW_USER_SIMD_LEVEL`:

```diff
diff --git a/docker-compose.yml b/docker-compose.yml
index 86cffad81..626292358 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -199,6 +199,7 @@ services:
       ARROW_JEMALLOC: "OFF"
       ARROW_S3: "OFF"
       ARROW_TEST_MEMCHECK: "ON"
+      ARROW_USER_SIMD_LEVEL: "avx2" # AVX512 not supported by Valgrind (ARROW-9851)
       ARROW_USE_LD_GOLD: "ON"
       BUILD_WARNING_LEVEL: "PRODUCTION"
     volumes: *conda-volumes
```

**Comment by pitrou** [comment]
> We need to use lower case for ARROW_USER_SIMD_LEVEL

Hmm, that's a pity. Can't we use the same convention everywhere?
(I would probably favour uppercase, since those names are generally abbreviations)

File: cpp/src/arrow/util/decimal_test.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/util/decimal_test.cc:395]: Done.

**Comment by kou** [comment]
I agree with you including upper case.

**Comment by jianxind** [comment]
I will create a JIRA for adding upper case to ARROW_USER_SIMD_LEVEL also.

**Comment by pitrou** [comment]
Since ARROW_USER_SIMD_LEVEL (environment variable) is enough to fix this, do we want to keep  ARROW_RUNTIME_SIMD_LEVEL (CMake setting) or not?

**Comment by pitrou** [comment]
@ursabot crossbow submit test-conda-cpp-valgrind

**Comment by jianxind** [comment]
> Since ARROW_USER_SIMD_LEVEL (environment variable) is enough to fix this, do we want to keep ARROW_RUNTIME_SIMD_LEVEL (CMake setting) or not?

I will vote +1 to keep ARROW_RUNTIME_SIMD_LEVEL, a  SIMD runtime level option at build time should benefit some cases.

For example, avx512 is disabled for MINGW build with a HACK. ARROW_RUNTIME_SIMD_LEVEL can be used here.
https://github.com/apache/arrow/blob/master/cpp/cmake_modules/SetupCxxFlags.cmake#L55

**Comment by kou** [comment]
OK.
I merge this.

**Pull request closed without merging**